src/swupd/fullfiles.go

```diff
@@ -0,0 +1,345 @@
+		if err == nil {
+			err = cerr
+		}
+	}()
+	err = tarRegularFullfile(uncompressed, input, name, fi)
+	if err != nil {
+		return fmt.Errorf("couldn't archive the file %s: %s", input, err)
+	}
+	uncompressedSize, err := uncompressed.Seek(0, io.SeekEnd)
+	if err != nil {
+		return fmt.Errorf("couldn't find the size of uncompressed fullfile: %s", err)
+	}
+
+	if DebugFullfiles {
+		log.Printf("DEBUG: Creating fullfile %s for regular file %s (%d bytes)", name, input, fi.Size())
+		log.Printf("DEBUG: %s (%d bytes, uncompressed)", filepath.Base(output), uncompressedSize)
+	}
+
+	// Pick the best compression option (or no compression) for that specific fullfile.
+	best := ""
+	var bestSize int64 = uncompressedSize
+	for i, c := range fullfileCompressors {
+		var candidateSize int64
+		candidate := fmt.Sprintf("%s.%d.%s", output, i, c.Name)
+
+		if c.Func != nil {
+			_, err := uncompressed.Seek(0, io.SeekStart)
+			if err != nil {
+				return fmt.Errorf("couldn't seek in fullfile %s: %s", input, err)
+			}
+			out, err := os.Create(candidate)
+			if err != nil {
+				log.Printf("WARNING: couldn't create output file for %q compressor: %s", c.Name, err)
+				continue
+			}
+			err = c.Func(out, uncompressed)
+			if err != nil {
+				log.Printf("WARNING: couldn't compress %s using compressor %q: %s", input, c.Name, err)
+				out.Close()
+				os.RemoveAll(candidate)
+				continue
+			}
+			candidateSize, err = out.Seek(0, io.SeekEnd)
+			if err != nil {
+				log.Printf("WARNING: couldn't get size of %s: %s", candidate, err)
+				out.Close()
+				os.RemoveAll(candidate)
+				continue
+			}
+			out.Close()
+		} else {
+			err = runTarCommand(input, name, candidate, c.ExternalTarExtraArg)
+			if err != nil {
+				log.Printf("WARNING: couldn't compress %s using compressor %q: %s", input, c.Name, err)
+				os.RemoveAll(candidate)
+				continue
+			}
+			candidateInfo, err := os.Stat(candidate)
+			if err != nil {
+				log.Printf("WARNING: couldn't get size of %s: %s", candidate, err)
+				os.RemoveAll(candidate)
+				continue
+			}
+			candidateSize = candidateInfo.Size()
+		}
+
+		if candidateSize < bestSize {
+			if best != "" {
+				os.RemoveAll(best)
+			}
+			best = candidate
+			bestSize = candidateSize
+		} else {
+			os.RemoveAll(candidate)
+		}
+
+		if DebugFullfiles {
+			log.Printf("DEBUG: %s (%d bytes)", filepath.Base(candidate), candidateSize)
+		}
+	}
+
+	if best != "" {
+		// Failure during rename might indicate some further problems, so return error
+		// instead of ignoring it (since there is an uncompressed version).
+		return os.Rename(best, output)
+	}
+	return nil
+}
+
+func runTarCommand(input, name, output, extra string) error {
+	var buf bytes.Buffer
+	// We know that name is suitable to be part of the transform regexp. If this function gets
+	// wider usage, we must revisit this.
+	cmd := exec.Command("tar", "--create", extra, "--file="+output, "--transform=s/.*/"+name+"/", "--preserve-permissions", input)
```

Comment:
You must tell tar to use numeric IDs, otherwise unpacking of a tar archive with a user that isn't defined on the client (yet, because it gets created as part of the update) will fail. Same for using the internal tar module.

I'm a bit worried that mixer now ends up creating the actual tar archive in two different ways: with the internal tar module or with the external tar command, just compressed differently. That's likely to give you slightly different results, which could become a source of future bugs. Wouldn't it be better to create the tar content internally, perhaps even buffered in memory, and then just compress it in different ways?

Has someone verified that the Go tar module handles xattrs correctly and in a way that is compatible with swupd-client, which now uses libarchive?

The goal was to not rely on external commands (see discussion around using mixer with Yocto when it runs under pseudo, aka https://github.com/clearlinux/swupd-server/pull/48). Shall I create an issue in this repo to track that?

Reply:
> You must tell tar to use numeric IDs, otherwise unpacking of a tar archive with a user that isn't defined on the client (yet, because it gets created as part of the update) will fail.

You mean passing --numeric-owner?

> Same for using the internal tar module.

The internal tar module should always include the numeric IDs.

> I'm a bit worried that mixer now ends up creating the actual tar archive in two different ways (...)

From the commit message:

    The original swupd-server uses tar and the compressors called from it,
    so keeping this around until we decide the Go's tar and compressors
    implementation is good to use in production.

I've also used to compare to see if the Go compressors was producing good enough file sizes. I should've added more clear comment about that. I plan to resolve this before this gets used by mixer.

> Has someone verified that the Go tar module handles xattrs correctly and in a way that is compatible with swupd-client, which now uses libarchive?

I haven't yet, but I've added this to my TODO list to check if the files we generate are good enough for swupd-client. Is there a list of specific xattrs you care about or you rely on propagating all of them?

> The goal was to not rely on external commands (...)

I'm moving toward that goal as well, I'll try to be more verbose in the commentary about that in the future.